src/utils.jl

```diff
@@ -208,3 +210,338 @@ end
 function typed_fsmi_fallback(fsmi::Type{T}, fargs...)::T where T
   return T()
 end
+
+mutable struct ScopeTracker
```

Comment:
`ScopeTracker` and all the new functionality below it is the heaviest work done in this PR, right? Could you sprinkle docstrings for these objects, just for future developpers (e.g. someone who in the future decides to rework this using JuliaLowering)? I am more curious about the "why" of this design than anything else.

Reply:
added some explanation